### ISSUE TITLE
Dealing with UDF types while integrating UDF service with the backend

### ISSUE THREAD

**Issue by pramitghosh** [Open-EO/openeo-udf#7]
Hi Soeren,

I have a preliminary draft implementation of a UDF service for R (currently only rasters) according to the specifications for the JSON structure for I/O to/from the UDF service that you had defined. However, while discussing the implementation with @flahn the other day, I found that there is no way for the UDF service to know which dimension is the UDF to be applied on. I mean to say, that suppose we have a simple UDF with an anonymous R function like this
```r
function(obj) {
  median(obj)
}
```
the UDF service will not know whether this is to be applied over time or over bands. Which dimension this is to be applied on is already pre-specified by the backend depending on which (native) process corresponding to one of the UDF types ("reduce_time", "apply_pixel" etc.) the UDF script is registered with. This information needs to be passed on to the UDF service somehow.

- So, I was thinking if it is possible to have that included in the "code" section of the JSON as a string such as `"time"`, `"band"` and so on. It could also be an array such as `["time", "band"]`.

- Another alternative would be to have different endpoints for the different UDF types in the UDF service which would then determine the dimension the function is to be applied on.

- One more alternative that came to my mind is to expose the object totally to the user. While this gives more flexibility to them it also brings much more responsibility with it. In this case, the user then needs to make sure whatever he/she does inside his/her UDF needs to conform to the native process in the backend that particular UDF is associated with.

What is your opinion on these and how do you address (or plan to address) this aspect of the integration of the UDF service in the Python world? I would personally prefer the 1st alternative since there everything is contained in the JSON itself and the UDF service would work the same way no matter which backend made the request. Said that, I am open to other alternatives too.

**Comment by m-mohr** [comment]
This issue could be broadened to the whole API. We currently have these processes called "min_time", "median_time" etc, which applies the min/median functions over time, but we also need to discuss whether that is really what we want there. Maybe we can agree on something consistent across the whole API?

Related to: https://github.com/Open-EO/openeo-api/issues/77 and https://github.com/Open-EO/openeo-api/issues/99

**Comment by pramitghosh** [comment]
Yes, it certainly could be @m-mohr. But it has to be kept in mind that in case of the UDFs the users define the functions *themselves* with full freedom and thus it becomes their responsibility to make sure their code does what it says in terms of the dimensionality (and otherwise things could go wrong).
